src/main/java/org/spongepowered/api/item/inventory/query/Query.java

```diff
@@ -24,26 +24,37 @@
  */
 package org.spongepowered.api.item.inventory.query;
 
-import org.spongepowered.api.CatalogType;
 import org.spongepowered.api.item.inventory.Inventory;
-import org.spongepowered.api.util.annotation.CatalogedBy;
 
 /**
- * Represents a possible type of operation for an {@linkplain
- * Inventory#query(QueryOperation...) inventory query}.
- *
- * @param <T> The argument type for the query
+ * An inventory query. See {@link QueryTypes} for possible types of queries.
  */
-@CatalogedBy(QueryOperationTypes.class)
-public interface QueryOperationType<T> extends CatalogType {
+public interface Query {
```

Comment:
This design to me is odd only because it has composite methods.

Generally you build a composite from a set of queries and have an adapter execute them. Here the query builds a composite in on itself (?) and then executes itself (??)

I think Inventory should get the execute method as you are basically asking for a view based on this query. I would then create utility methods for common composites (and, not, or) and pass Query objects to them.

Reply:
The logic for what to do is in the `Query` impl. It cannot be in the Inventory/Adapter.
`Inventory` has `query(QueryType, param(s))` and `query(Query)`  which (creates) + executes the query with itself as input.

Reply:
Which is fine, in the implementation you could keep the logic in Query. I just would rather not see that at the API level.

From the API's perspective, Query should be kept dumb. There should be methods to build composite queries. Inventory should have the execute method that takes the Query. In the impl, if you want to have Query do the logic and basically forward the Inventory call to it then knock yourself out. At least the API would remain clean in that regard.